src/main/java/com/ldtteam/structures/client/BlueprintHandler.java

```diff
@@ -62,25 +52,53 @@ public static BlueprintHandler getInstance()
      * @param blueprint the wayPointBlueprint to draw.
      * @param pos       its position.
      */
-    public void draw(
-        final Blueprint blueprint,
-        final BlockPos pos,
-        final MatrixStack stack,
-        final float partialTicks)
+    public void draw(final Blueprint blueprint, final BlockPos pos, final MatrixStack stack, final float partialTicks)
     {
         if (blueprint == null)
         {
             Log.getLogger().warn("Trying to draw null blueprint!");
             return;
         }
+        Minecraft.getInstance().getProfiler().startSection("struct_render_cache");
 
-        try
+        final int blueprintHash = blueprint.hashCode();
+        final SoftReference<BlueprintRenderer> rendererRef = rendererCache.get(blueprintHash);
+        final BlueprintRenderer renderer;
+
+        if (rendererRef == null || rendererRef.get() == null)
+        {
+            renderer = BlueprintRenderer.buildRendererForBlueprint(blueprint);
+            rendererCache.put(blueprintHash, new SoftReference<>(renderer));
+        }
+        else
         {
-            blueprintBufferBuilderCache.get(blueprint, () -> BlueprintRenderer.buildRendererForBlueprint(blueprint)).draw(pos, stack, partialTicks);
+            renderer = rendererRef.get();
         }
-        catch (ExecutionException e)
+
+        evictTimeCache.put(blueprintHash, System.currentTimeMillis());
+        renderer.updateBlueprint(blueprint);
+        renderer.draw(pos, stack, partialTicks);
+
+        Minecraft.getInstance().getProfiler().endSection();
+    }
+
+    /**
+     * Cleans entries that are older than CACHE_EVICT_TIME.
+     */
+    public void cleanCache()
+    {
+        for (final Int2LongMap.Entry entry : evictTimeCache.int2LongEntrySet())
         {
-            Log.getLogger().error(e);
+            if (entry.getLongValue() + CACHE_EVICT_TIME < System.currentTimeMillis())
+            {
+                final int removeHash = entry.getIntKey();
+                final SoftReference<BlueprintRenderer> removeRendererRef = rendererCache.remove(removeHash);
+                evictTimeCache.remove(removeHash);
```

Comment:
can you actually remove it during iteration here without an exception?

Reply:
its ArrayMap, and it never threw an exception during testing (was logging it so it surely happened many times)

Reply:
btw the for each uses set method which creates a copy, so that's probably why